app/src/main/java/io/mochahub/powermeter/workoutsession/workoutsessiondialog/NewWorkoutDialog.kt

```diff
@@ -0,0 +1,340 @@
+        }
+
+        val myCalendar = Calendar.getInstance()
+        val dateSetListener = DatePickerDialog.OnDateSetListener {
+                _, year, month, day -> myCalendar.apply {
+            set(Calendar.YEAR, year)
+            set(Calendar.MONTH, month)
+            set(Calendar.DAY_OF_MONTH, day)
+            newWorkoutDateText.setText(viewModel.simpleDateFormat.format(myCalendar.time))
+        }
+        }
+
+        newWorkoutDateText.setOnClickListener {
+            DatePickerDialog(requireContext(), dateSetListener, myCalendar
+                .get(Calendar.YEAR), myCalendar.get(Calendar.MONTH),
+                myCalendar.get(Calendar.DAY_OF_MONTH)).show()
+        }
+    }
+
+    // //////////////////////////////////////////////////////////////
+    // Interface methods for workout controller
+    // //////////////////////////////////////////////////////////////
+    // Need to use a loop method because it is more reliable then indexing
+    // If we use index's there is a bug when deleting sets/ workouts
+    // that causes out of bounds exceptions
+    override fun onAddSetClicked(workout: Workout) {
+        for (i in 0 until viewModel.workouts.size) {
+            if (viewModel.workouts[i].id == workout.id) {
+                viewModel.workouts[i] = viewModel.workouts[i].addSet(0, WorkoutSet(weight = 0.0, reps = 0))
+                workoutController.setData(viewModel.workouts)
+            }
+        }
+    }
+
+    override fun onRepFocusChanged(workout: Workout, workoutSet: WorkoutSet, value: Int) {
+        for (i in 0 until viewModel.workouts.size) {
+            if (viewModel.workouts[i].id == workout.id) {
+                for (k in 0 until viewModel.workouts[i].sets.size) {
+                    if (viewModel.workouts[i].sets[k].id == workoutSet.id) {
+                        viewModel.workouts[i].sets[k] = viewModel.workouts[i].sets[k].setReps(value)
+                        workoutController.setData(viewModel.workouts)
+                    }
+                }
+            }
+        }
+    }
+
+    override fun onWeightFocusChanged(workout: Workout, workoutSet: WorkoutSet, value: Double) {
+        for (i in 0 until viewModel.workouts.size) {
+            if (viewModel.workouts[i].id == workout.id) {
+                for (k in 0 until viewModel.workouts[i].sets.size) {
+                    if (viewModel.workouts[i].sets[k].id == workoutSet.id) {
+                        viewModel.workouts[i].sets[k] = viewModel.workouts[i].sets[k].setWeight(value)
+                        workoutController.setData(viewModel.workouts)
+                    }
+                }
+            }
+        }
+    }
+
+    override fun onExerciseSelected(workout: Workout, exercise: String) {
+
+        val foundExercise = exercises.find { it.name == exercise }
+        if (foundExercise == null) {
+            Log.e(this.javaClass.canonicalName, "Exercise not found")
+            return
+        }
```

Comment:
This state should never occur right? I'd get rid of this check, I'd rather crash when this happens so we can find the error

Reply:
Fixed here https://github.com/power-meter/android-client/pull/32/commits/132b5a1788d05681550cf40195c8d4471af37d57